From: Konsta <konsta09@gmail.com>
Subject: framework/base (0001): add qualcomm fm support

diff --git a/core/java/android/os/Process.java b/core/java/android/os/Process.java
index aaae3c7..fc00cc5 100644
--- a/core/java/android/os/Process.java
+++ b/core/java/android/os/Process.java
@@ -109,6 +109,12 @@ public class Process {
      */
     public static final int VPN_UID = 1016;
 
+     /**
+     * Defines the UID/GID for the FM process.
+     * @hide
+     */
+    public static final int FM_RADIO_UID = 1033;
+
     /**
      * Defines the UID/GID for the NFC service process.
      * @hide
diff --git a/data/etc/platform.xml b/data/etc/platform.xml
index 47cb7ab..7326c92 100644
--- a/data/etc/platform.xml
+++ b/data/etc/platform.xml
@@ -142,5 +142,7 @@
             file="/system/framework/android.test.runner.jar" />
     <library name="javax.obex"
             file="/system/framework/javax.obex.jar"/>
+    <library name="qcom.fmradio"
+            file="/system/framework/qcom.fmradio.jar"/>
 
 </permissions>
diff --git a/services/java/com/android/server/pm/PackageManagerService.java b/services/java/com/android/server/pm/PackageManagerService.java
index 09d02cd..976d71f 100644
--- a/services/java/com/android/server/pm/PackageManagerService.java
+++ b/services/java/com/android/server/pm/PackageManagerService.java
@@ -193,11 +193,16 @@ public class PackageManagerService extends IPackageManager.Stub {
     private static final boolean DEBUG_APP_DIR_OBSERVER = false;
     private static final boolean DEBUG_VERIFY = false;
 
+    static final boolean MULTIPLE_APPLICATION_UIDS = true;
     private static final int RADIO_UID = Process.PHONE_UID;
     private static final int LOG_UID = Process.LOG_UID;
     private static final int NFC_UID = Process.NFC_UID;
     private static final int BLUETOOTH_UID = Process.BLUETOOTH_UID;
     private static final int SHELL_UID = Process.SHELL_UID;
+    private static final int FM_RADIO_UID = Process.FM_RADIO_UID;
+    static final int FIRST_APPLICATION_UID =
+                   Process.FIRST_APPLICATION_UID;
+    static final int MAX_APPLICATION_UIDS = 1000;
 
     private static final boolean GET_CERTIFICATES = true;
 
@@ -1092,6 +1097,8 @@ public class PackageManagerService extends IPackageManager.Stub {
         mSettings.addSharedUserLPw("android.uid.nfc", NFC_UID, ApplicationInfo.FLAG_SYSTEM);
         mSettings.addSharedUserLPw("android.uid.bluetooth", BLUETOOTH_UID, ApplicationInfo.FLAG_SYSTEM);
         mSettings.addSharedUserLPw("android.uid.shell", SHELL_UID, ApplicationInfo.FLAG_SYSTEM);
+        mSettings.addSharedUserLPw("android.uid.fm_radio", MULTIPLE_APPLICATION_UIDS
+                ? FM_RADIO_UID : FIRST_APPLICATION_UID, ApplicationInfo.FLAG_SYSTEM);
 
         mAppOps = (AppOpsManager) context.getSystemService(Context.APP_OPS_SERVICE);
         String separateProcesses = SystemProperties.get("debug.separate_processes");
-- 
1.7.9.5

