From: ChepKun <ProgrammerKun@gmail.com>
Subject: frameworks/base (9999): tempory fix msim

diff --git a/packages/SystemUI/src/com/android/systemui/statusbar/phone/PhoneStatusBar.java b/packages/SystemUI/src/com/android/systemui/statusbar/phone/PhoneStatusBar.java
index 6b9c37f..96ecf79 100644
--- a/packages/SystemUI/src/com/android/systemui/statusbar/phone/PhoneStatusBar.java
+++ b/packages/SystemUI/src/com/android/systemui/statusbar/phone/PhoneStatusBar.java
@@ -488,13 +488,13 @@ public class PhoneStatusBar extends BaseStatusBar {
 
         mIconSize = res.getDimensionPixelSize(com.android.internal.R.dimen.status_bar_icon_size);
 
-        if (MSimTelephonyManager.getDefault().isMultiSimEnabled()) {
+       /* if (MSimTelephonyManager.getDefault().isMultiSimEnabled()) {
             mStatusBarWindow = (StatusBarWindowView) View.inflate(context,
                     R.layout.msim_super_status_bar, null);
-        } else {
+        } else {*/
             mStatusBarWindow = (StatusBarWindowView) View.inflate(context,
                     R.layout.super_status_bar, null);
-        }
+        //}
         mStatusBarWindow.mService = this;
         mStatusBarWindow.setOnTouchListener(new View.OnTouchListener() {
             @Override
@@ -507,21 +507,21 @@ public class PhoneStatusBar extends BaseStatusBar {
                 return mStatusBarWindow.onTouchEvent(event);
             }});
 
-        if (MSimTelephonyManager.getDefault().isMultiSimEnabled()) {
+        /*if (MSimTelephonyManager.getDefault().isMultiSimEnabled()) {
             mStatusBarView = (PhoneStatusBarView) mStatusBarWindow.findViewById(
                     R.id.msim_status_bar);
-        } else {
+        } else {*/
             mStatusBarView = (PhoneStatusBarView) mStatusBarWindow.findViewById(R.id.status_bar);
-        }
+        //}
         mStatusBarView.setBar(this);
 
 
         PanelHolder holder;
-        if (MSimTelephonyManager.getDefault().isMultiSimEnabled()) {
+       /* if (MSimTelephonyManager.getDefault().isMultiSimEnabled()) {
             holder = (PanelHolder) mStatusBarWindow.findViewById(R.id.msim_panel_holder);
-        } else {
+        } else {*/
             holder = (PanelHolder) mStatusBarWindow.findViewById(R.id.panel_holder);
-        }
+       // }
         mStatusBarView.setPanelHolder(holder);
 
         mNotificationPanel = (NotificationPanelView) mStatusBarWindow.findViewById(R.id.notification_panel);
@@ -682,7 +682,7 @@ public class PhoneStatusBar extends BaseStatusBar {
 
         mSignalView.setNetworkController(mNetworkController);
 
-        if (MSimTelephonyManager.getDefault().isMultiSimEnabled()) {
+        /*if (MSimTelephonyManager.getDefault().isMultiSimEnabled()) {
             mMSimNetworkController = new MSimNetworkController(mContext);
             MSimSignalClusterView mSimSignalCluster = (MSimSignalClusterView)
               mStatusBarView.findViewById(R.id.msim_signal_cluster);
@@ -727,7 +727,7 @@ public class PhoneStatusBar extends BaseStatusBar {
                     }
                 });
             }
-        } else {
+        } else {*/
             mNetworkController = new NetworkController(mContext);
             final SignalClusterView signalCluster =
                 (SignalClusterView)mStatusBarView.findViewById(R.id.signal_cluster);
@@ -776,7 +776,7 @@ public class PhoneStatusBar extends BaseStatusBar {
                     }
                 });
             }
-        }
+        //}
 
         // Quick Settings (where available, some restrictions apply)
         if (mHasSettingsPanel) {
@@ -1360,17 +1360,17 @@ public class PhoneStatusBar extends BaseStatusBar {
 
         final boolean emergencyCallsShownElsewhere = mEmergencyCallLabel != null;
         final boolean makeVisible;
-        if (MSimTelephonyManager.getDefault().isMultiSimEnabled()) {
+        /*if (MSimTelephonyManager.getDefault().isMultiSimEnabled()) {
             makeVisible =
                 !(emergencyCallsShownElsewhere && mMSimNetworkController.isEmergencyOnly())
                 && mPile.getHeight() < (mNotificationPanel.getHeight() - mCarrierLabelHeight
                  - mNotificationHeaderHeight) && mScrollView.getVisibility() == View.VISIBLE;
-        } else {
+        } else {*/
             makeVisible =
                 !(emergencyCallsShownElsewhere && mNetworkController.isEmergencyOnly())
                 && mPile.getHeight() < (mNotificationPanel.getHeight() - mCarrierLabelHeight
                   - mNotificationHeaderHeight)&& mScrollView.getVisibility() == View.VISIBLE;
-        }
+        //}
 
         if (force || mCarrierLabelVisible != makeVisible) {
             mCarrierLabelVisible = makeVisible;
@@ -2335,11 +2335,11 @@ public class PhoneStatusBar extends BaseStatusBar {
             final View notifications = mStatusBarView.findViewById(R.id.notification_icon_area);
             final View systemIcons = mStatusBarView.findViewById(R.id.statusIcons);
             final View signal;
-            if (MSimTelephonyManager.getDefault().isMultiSimEnabled()) {
+            /*if (MSimTelephonyManager.getDefault().isMultiSimEnabled()) {
                 signal = mStatusBarView.findViewById(R.id.msim_signal_cluster);
-            } else {
+            } else {*/
                 signal = mStatusBarView.findViewById(R.id.signal_cluster);
-            }
+            //}
             final View signal2 = mStatusBarView.findViewById(R.id.signal_cluster_text);
             final View battery = mStatusBarView.findViewById(R.id.battery);
             final View battery2 = mStatusBarView.findViewById(R.id.battery_text);
@@ -2629,13 +2629,13 @@ public class PhoneStatusBar extends BaseStatusBar {
             mGestureRec.dump(fd, pw, args);
         }
 
-        if (MSimTelephonyManager.getDefault().isMultiSimEnabled()) {
+        /*if (MSimTelephonyManager.getDefault().isMultiSimEnabled()) {
             for(int i=0; i < MSimTelephonyManager.getDefault().getPhoneCount(); i++) {
                 mMSimNetworkController.dump(fd, pw, args, i);
             }
-        } else {
+        } else {*/
             mNetworkController.dump(fd, pw, args);
-        }
+        //}
     }
 
     @Override
diff --git a/packages/SystemUI/src/com/android/systemui/statusbar/policy/MSimNetworkController.java b/packages/SystemUI/src/com/android/systemui/statusbar/policy/MSimNetworkController.java
index f950c71..8b7a753 100644
--- a/packages/SystemUI/src/com/android/systemui/statusbar/policy/MSimNetworkController.java
+++ b/packages/SystemUI/src/com/android/systemui/statusbar/policy/MSimNetworkController.java
@@ -58,7 +58,7 @@ import com.android.systemui.R;
 public class MSimNetworkController extends NetworkController {
     // debug
     static final String TAG = "StatusBar.MSimNetworkController";
-    static final boolean DEBUG = false;
+    static final boolean DEBUG = true;
     static final boolean CHATTY = true; // additional diagnostics, but not logspew
 
     // telephony
