From: ChepKun <ProgrammerKun@gmail.com>
Subject: bootable/recovery (0001): add buttons backlight control

diff --git a/Android.mk b/Android.mk
index ff1f5a2..3b4b7f2 100644
--- a/Android.mk
+++ b/Android.mk
@@ -251,6 +251,9 @@ endif
 ifneq ($(TW_BRIGHTNESS_PATH),)
 	LOCAL_CFLAGS += -DTW_BRIGHTNESS_PATH=$(TW_BRIGHTNESS_PATH)
 endif
+ifneq ($(TW_BUTTONS_BRIGHTNESS_PATH),)
+	LOCAL_CFLAGS += -DTW_BUTTONS_BRIGHTNESS_PATH=$(TW_BUTTONS_BRIGHTNESS_PATH)
+endif
 ifneq ($(TW_MAX_BRIGHTNESS),)
 	LOCAL_CFLAGS += -DTW_MAX_BRIGHTNESS=$(TW_MAX_BRIGHTNESS)
 endif
diff --git a/data.cpp b/data.cpp
index c611e78..5f34d30 100644
--- a/data.cpp
+++ b/data.cpp
@@ -939,6 +939,9 @@ void DataManager::SetDefaultValues()
 		LOGINFO("TW_BRIGHTNESS_PATH := %s\n", EXPAND(TW_BRIGHTNESS_PATH));
 		mConstValues.insert(make_pair("tw_has_brightnesss_file", "1"));
 		mConstValues.insert(make_pair("tw_brightness_file", EXPAND(TW_BRIGHTNESS_PATH)));
+	#ifdef TW_BUTTONS_BRIGHTNESS_PATH
+		mConstValues.insert(make_pair("tw_buttons_brightness_file", EXPAND(TW_BUTTONS_BRIGHTNESS_PATH)));
+	#endif
 		ostringstream maxVal;
 		maxVal << TW_MAX_BRIGHTNESS;
 		mConstValues.insert(make_pair("tw_brightness_max", maxVal.str()));
@@ -1130,6 +1133,10 @@ void DataManager::ReadSettingsFile(void)
 		string brightness_path = EXPAND(TW_BRIGHTNESS_PATH);
 		string brightness_value = GetStrValue("tw_brightness");
 		TWFunc::write_file(brightness_path, brightness_value);
+#ifdef TW_BUTTONS_BRIGHTNESS_PATH
+		string btns_brightness_path = EXPAND(TW_BUTTONS_BRIGHTNESS_PATH);
+		TWFunc::write_file(btns_brightness_path, brightness_value);
+#endif
 	}
 #endif
 }
diff --git a/gui/devices/480x854/res/ui.xml b/gui/devices/480x854/res/ui.xml
index 08c3ddc..32cc215 100644
--- a/gui/devices/480x854/res/ui.xml
+++ b/gui/devices/480x854/res/ui.xml
@@ -2633,6 +2633,7 @@
 					<action function="compute">tw_brightness*%tw_brightness_pct%</action>
 					<action function="compute">tw_brightness/100</action>
 					<action function="cmd">echo %tw_brightness% > &quot;%tw_brightness_file%&quot;</action>
+					<action function="cmd">echo %tw_brightness% > &quot;%tw_buttons_brightness_file%&quot;</action>
 				</actions>
 			</object>
 
