From: ChepKun <ProgrammerKun@gmail.com>
Subject: build (0002): fix progress bar in recovery 

diff --git a/tools/releasetools/ota_from_target_files b/tools/releasetools/ota_from_target_files
index d0f8b3c..fe6eca9 100755
--- a/tools/releasetools/ota_from_target_files
+++ b/tools/releasetools/ota_from_target_files
@@ -509,6 +509,8 @@ reboot_now("%(bcb_dev)s", "recovery");
 else if get_stage("%(bcb_dev)s", "stage") == "3/3" then
 """ % bcb_dev)
 
+  script.ShowProgress(0.5, 20)
+
   device_specific.FullOTA_InstallBegin()
 
   script.Mount("/data")
@@ -520,8 +522,6 @@ else if get_stage("%(bcb_dev)s", "stage") == "3/3" then
     script.RunBackup("backup")
     script.Unmount("/system")
 
-  script.ShowProgress(0.5, 0)
-
   if OPTIONS.wipe_user_data:
     script.FormatPartition("/data")
 
@@ -547,18 +547,17 @@ else if get_stage("%(bcb_dev)s", "stage") == "3/3" then
 
   common.CheckSize(boot_img.data, "boot.img", OPTIONS.info_dict)
   common.ZipWriteStr(output_zip, "boot.img", boot_img.data)
-  script.ShowProgress(0.2, 0)
 
   script.ValidateSignatures("cleanup")
 
   if OPTIONS.backuptool:
-    script.ShowProgress(0.2, 10)
+    script.ShowProgress(0.3, 10)
     script.RunBackup("restore")
 
-  script.ShowProgress(0.2, 10)
+  script.ShowProgress(0.2, 5)
   script.WriteRawImage("/boot", "boot.img")
 
-  script.ShowProgress(0.1, 0)
+  script.ShowProgress(0.1, 5)
   device_specific.FullOTA_InstallEnd()
 
   if OPTIONS.extra_script is not None:
