From: ChepKun <ProgrammerKun@gmail.com>
Subject: packages/Settings (0001): don't hide high end gfx settings

diff --git a/src/com/android/settings/cyanogenmod/PerformanceSettings.java b/src/com/android/settings/cyanogenmod/PerformanceSettings.java
index 1c4d413..c01be72 100644
--- a/src/com/android/settings/cyanogenmod/PerformanceSettings.java
+++ b/src/com/android/settings/cyanogenmod/PerformanceSettings.java
@@ -106,13 +106,9 @@ public class PerformanceSettings extends SettingsPreferenceFragment implements
         String use16bppAlpha = SystemProperties.get(USE_16BPP_ALPHA_PROP, "0");
         mUse16bppAlphaPref.setChecked("1".equals(use16bppAlpha));
 
-        if (ActivityManager.isLowRamDeviceStatic()) {
-            mForceHighEndGfx = (CheckBoxPreference) prefSet.findPreference(FORCE_HIGHEND_GFX_PREF);
-            String forceHighendGfx = SystemProperties.get(FORCE_HIGHEND_GFX_PERSIST_PROP, "false");
-            mForceHighEndGfx.setChecked("true".equals(forceHighendGfx));
-        } else {
-            prefSet.removePreference(findPreference(FORCE_HIGHEND_GFX_PREF));
-        }
+        mForceHighEndGfx = (CheckBoxPreference) prefSet.findPreference(FORCE_HIGHEND_GFX_PREF);
+        String forceHighendGfx = SystemProperties.get(FORCE_HIGHEND_GFX_PERSIST_PROP, "true");
+        mForceHighEndGfx.setChecked("true".equals(forceHighendGfx));
 
         /* Display the warning dialog */
         alertDialog = new AlertDialog.Builder(getActivity()).create();
