From: ChepKun <ProgrammerKun@gmail.com>
Subject: packages/Settings (0002): don't hide performance settings

diff --git a/src/com/android/settings/Settings.java b/src/com/android/settings/Settings.java
index 8b77fcd..ba87954 100644
--- a/src/com/android/settings/Settings.java
+++ b/src/com/android/settings/Settings.java
@@ -658,8 +658,7 @@ public class Settings extends PreferenceActivity
                         target.remove(i);
                     }
                 }
-            } else if (id == R.id.development_settings
-                    || id == R.id.performance_settings) {
+            } else if (id == R.id.development_settings) {
                 if (!showDev) {
                     target.remove(i);
                 }
