From: Konsta <konsta09@gmail.com>
Subject: packages/Settings (0005_1): fix USB charging

diff --git a/src/com/android/settings/cyanogenmod/MoreDeviceSettings.java b/src/com/android/settings/cyanogenmod/MoreDeviceSettings.java
index 9adaeef..f3e29dc 100644
--- a/src/com/android/settings/cyanogenmod/MoreDeviceSettings.java
+++ b/src/com/android/settings/cyanogenmod/MoreDeviceSettings.java
@@ -45,6 +45,15 @@ public class MoreDeviceSettings extends SettingsPreferenceFragment {
 
 	private CheckBoxPreference mUsbCharging;
 
+    private static boolean isUsbChargingSupported() {
+        try {
+            return UsbCharging.isSupported();
+        } catch (NoClassDefFoundError e) {
+            // Hardware abstraction framework not installed
+            return false;
+        }
+    }
+
     @Override
     public void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
@@ -102,14 +111,7 @@ public class MoreDeviceSettings extends SettingsPreferenceFragment {
             final SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
             final boolean enabled = prefs.getBoolean(KEY_USB_CHARGING, true);
             UsbCharging.setEnabled(enabled);
-    }
+    	}
+	}
 
-    private static boolean isUsbChargingSupported() {
-        try {
-            return UsbCharging.isSupported();
-        } catch (NoClassDefFoundError e) {
-            // Hardware abstraction framework not installed
-            return false;
-        }
-    }
 }
