From: ChepKun <ProgrammerKun@gmail.com>
Subject: packages/Settings (0006): add option to enable navbar

diff --git a/AndroidManifest.xml b/AndroidManifest.xml
index db925a5..2d4ce02 100644
--- a/AndroidManifest.xml
+++ b/AndroidManifest.xml
@@ -66,6 +66,7 @@
     <uses-permission android:name="android.permission.SET_TIME" />
     <uses-permission android:name="android.permission.ACCESS_NOTIFICATIONS" />
     <uses-permission android:name="android.permission.CHANGE_PRIVACY_GUARD_STATE" />
+	<uses-permission android:name="android.permission.REBOOT"/>
 
     <permission
         android:name="android.permission.REQUEST_SUPERUSER"
diff --git a/res/values-ru/cm_strings.xml b/res/values-ru/cm_strings.xml
index dcfd4f1..fd8e30f 100644
--- a/res/values-ru/cm_strings.xml
+++ b/res/values-ru/cm_strings.xml
@@ -22,6 +22,9 @@
     <string name="navigation_bar_height">Высота кнопок</string>
     <string name="navigation_bar_height_48">48dp (по умолчанию)</string>
 
+    <!-- Navbar -->
+    <string name="show_navbar">Строка навигации</string>
+    <string name="show_navbar_summary">Требуется перезагрузка</string>
     <!-- Quiet hours -->
     <string name="quiet_hours_title">Тихие часы</string>
     <string name="quiet_hours_summary">Настройка времени, когда устройство должно быть тихим</string>
diff --git a/res/values/cm_strings.xml b/res/values/cm_strings.xml
index 4ce1eaf..48244ac 100644
--- a/res/values/cm_strings.xml
+++ b/res/values/cm_strings.xml
@@ -29,6 +29,10 @@
     <string name="navigation_bar_height_28">28dp</string>
     <string name="navigation_bar_height_24">24dp</string>
 
+    <!-- Navbar -->
+    <string name="show_navbar">Navigation bar</string>
+    <string name="show_navbar_summary">Requires reboot</string>
+
     <!-- Activity label of BluetoothMasPermissionActivity, also used as Strings in the permission dialog [CHAR LIMIT=none] -->
     <string name="bluetooth_mas_request">"Message Access request"</string>
     <!-- Bluetooth MAS permission Alert Activity text [CHAR LIMIT=none] -->
diff --git a/res/xml/system_ui_settings.xml b/res/xml/system_ui_settings.xml
index b91765a..77a9c54 100644
--- a/res/xml/system_ui_settings.xml
+++ b/res/xml/system_ui_settings.xml
@@ -51,15 +51,21 @@
         android:fragment="com.android.settings.cyanogenmod.PieControl"
         android:title="@string/pie_control_title" />
 
-    <PreferenceScreen
-        android:key="navigation_bar"
-        android:fragment="com.android.settings.cyanogenmod.NavBar"
-        android:title="@string/navigation_bar_title" />
+    <CheckBoxPreference
+        android:key="show_navbar"
+        android:title="@string/show_navbar"
+        android:summary="@string/show_navbar_summary"
+        android:defaultValue="false" />
 
     <PreferenceCategory
         android:key="navigation_bar"
         android:title="@string/navigation_bar_category" >
 
+    	<PreferenceScreen
+        	android:key="navigation_bar"
+        	android:fragment="com.android.settings.cyanogenmod.NavBar"
+        	android:title="@string/navigation_bar_title" />
+
         <ListPreference
             android:key="nav_buttons_height"
             android:dialogTitle="@string/navigation_bar_height"
diff --git a/src/com/android/settings/cyanogenmod/SystemUiSettings.java b/src/com/android/settings/cyanogenmod/SystemUiSettings.java
index 39acd65..85b6aee 100644
--- a/src/com/android/settings/cyanogenmod/SystemUiSettings.java
+++ b/src/com/android/settings/cyanogenmod/SystemUiSettings.java
@@ -16,9 +16,13 @@
 
 package com.android.settings.cyanogenmod;
 
+import android.app.AlertDialog;
 import android.content.ContentResolver;
+import android.content.Context;
+import android.content.DialogInterface;
 import android.content.res.Resources;
 import android.os.Bundle;
+import android.os.PowerManager;
 import android.os.RemoteException;
 import android.preference.CheckBoxPreference;
 import android.preference.ListPreference;
@@ -39,6 +43,7 @@ public class SystemUiSettings extends SettingsPreferenceFragment  implements
     private static final String KEY_EXPANDED_DESKTOP = "expanded_desktop";
     private static final String KEY_EXPANDED_DESKTOP_NO_NAVBAR = "expanded_desktop_no_navbar";
     private static final String CATEGORY_NAVBAR = "navigation_bar";
+	private static final String KEY_SHOW_NAVBAR = "show_navbar";
 	private static final String KEY_NAVIGATION_HEIGHT = "nav_buttons_height";
     private static final String KEY_PIE_CONTROL = "pie_control";
     private static final String KEY_SCREEN_GESTURE_SETTINGS = "touch_screen_gesture_settings";
@@ -47,6 +52,7 @@ public class SystemUiSettings extends SettingsPreferenceFragment  implements
     private ListPreference mExpandedDesktopPref;
     private CheckBoxPreference mExpandedDesktopNoNavbarPref;
     private ListPreference mNavButtonsHeight;
+	private CheckBoxPreference mShowNavbar;
 
     @Override
     public void onCreate(Bundle savedInstanceState) {
@@ -68,6 +74,11 @@ public class SystemUiSettings extends SettingsPreferenceFragment  implements
         int expandedDesktopValue = Settings.System.getInt(getContentResolver(),
                 Settings.System.EXPANDED_DESKTOP_STYLE, 0);
 
+		// Show navigation bar
+        mShowNavbar = (CheckBoxPreference) findPreference(KEY_SHOW_NAVBAR);
+        mShowNavbar.setChecked(Settings.System.getInt(getContentResolver(),
+                Settings.System.SHOW_NAVBAR, 0) == 1);
+
         // Navbar height
         mNavButtonsHeight = (ListPreference) findPreference(KEY_NAVIGATION_HEIGHT);
         if (mNavButtonsHeight != null) {
@@ -167,4 +178,16 @@ public class SystemUiSettings extends SettingsPreferenceFragment  implements
             mExpandedDesktopPref.setSummary(res.getString(summary));
         }
     }
+
+ 	public boolean onPreferenceTreeClick(PreferenceScreen preferenceScreen, Preference preference) {
+		// Show or hide navigation bar
+        if (preference == mShowNavbar) {
+            Settings.System.putInt(getContentResolver(), Settings.System.SHOW_NAVBAR,
+                    mShowNavbar.isChecked() ? 1 : 0);	
+        } else {
+           return super.onPreferenceTreeClick(preferenceScreen, preference);
+        }
+        return true;
+    }
+
 }
