From: ChepKun <ProgrammerKun@gmail.com>
Subject: packages/Settings (0008): don't hide performance settings

diff --git a/src/com/android/settings/Settings.java b/src/com/android/settings/Settings.java
index 8c139ed..27e8ec3 100644
--- a/src/com/android/settings/Settings.java
+++ b/src/com/android/settings/Settings.java
@@ -501,8 +501,7 @@ public class Settings extends PreferenceActivity
                         || Utils.isMonkeyRunning()) {
                     target.remove(i);
                 }
-            } else if (id == R.id.development_settings
-                    || id == R.id.performance_settings) {
+	    } else if (id == R.id.development_settings) {
                 if (!showDev) {
                     target.remove(i);
                 }
