From: ChepKun <ProgrammerKun@gmail.com>
Subject: external/noto-fonts (0001): build noto-fonts with SMALLER_FONT_FOOTPRINT flag

diff --git a/Android.mk b/Android.mk
index 4bc06fc..9911899 100644
--- a/Android.mk
+++ b/Android.mk
@@ -17,8 +17,6 @@
 
 LOCAL_PATH := $(call my-dir)
 
-ifneq ($(SMALLER_FONT_FOOTPRINT),true)
-
 include $(CLEAR_VARS)
 LOCAL_MODULE := NotoColorEmoji.ttf
 LOCAL_SRC_FILES := $(LOCAL_MODULE)
@@ -322,5 +320,3 @@ LOCAL_MODULE_CLASS := ETC
 LOCAL_MODULE_TAGS := optional
 LOCAL_MODULE_PATH := $(TARGET_OUT)/fonts
 include $(BUILD_PREBUILT)
-
-endif
