From: ChepKun <ProgrammerKun@gmail.com>
Subject: vendor/cm (0001): exclude some packages

diff --git a/config/common_full.mk b/config/common_full.mk
index dc2e40d..9b00d2c 100644
--- a/config/common_full.mk
+++ b/config/common_full.mk
@@ -9,14 +9,7 @@ PRODUCT_PACKAGE_OVERLAYS += vendor/cm/overlay/dictionaries
 
 # Optional CM packages
 PRODUCT_PACKAGES += \
-    Galaxy4 \
-    HoloSpiralWallpaper \
-    LiveWallpapers \
     LiveWallpapersPicker \
-    MagicSmokeWallpapers \
-    NoiseField \
-    PhaseBeam \
-    VisualizationWallpapers \
     PhotoTable \
     SoundRecorder
 
diff --git a/config/telephony.mk b/config/telephony.mk
index 1c0c4a6..097693f 100644
--- a/config/telephony.mk
+++ b/config/telephony.mk
@@ -14,7 +14,6 @@ PRODUCT_COPY_FILES += \
 # Telephony packages
 PRODUCT_PACKAGES += \
     Mms \
-    Stk \
     CellBroadcastReceiver \
     VoiceDialer \
     WhisperPush
