From: ChepKun <ProgrammerKun@gmail.com>
Subject: vendor/cm (0002): add OTA

diff --git a/config/common.mk b/config/common.mk
index 450eb10..4bdb3e8 100644
--- a/config/common.mk
+++ b/config/common.mk
@@ -134,14 +134,12 @@ PRODUCT_PACKAGES += \
     DSPManager \
     libcyanogen-dsp \
     audio_effects.conf \
-    CMWallpapers \
     Apollo \
     CMFileManager \
     LockClock \
-    CMUpdater \
-    CMFota \
     CMAccount \
-    CMHome
+    CMHome \
+    OTAUpdateCenter
 
 # CM Hardware Abstraction Framework
 PRODUCT_PACKAGES += \
@@ -324,6 +322,15 @@ PRODUCT_PROPERTY_OVERRIDES += persist.sys.recovery_update=false
 PRODUCT_PROPERTY_OVERRIDES += \
   ro.cm.display.version=$(CM_DISPLAY_VERSION)
 
+# OTA
+OTA_PRODUCT_NAME := cm
+OTA_DATE_TIME := $(shell date -u +%Y%m%d-%H%M)
+
+PRODUCT_PROPERTY_OVERRIDES += \
+      otaupdater.otaid=$(OTA_DEVELOPER_ID)-$(OTA_PRODUCT_NAME)$(PRODUCT_VERSION_MAJOR)-$(CM_BUILD) \
+      otaupdater.otaver=$(OTA_PRODUCT_NAME)-$(CM_VERSION) \
+      otaupdater.otatime=$(OTA_DATE_TIME)
+
 # disable multithreaded dextop for RELEASE and SNAPSHOT builds
 ifneq ($(filter RELEASE SNAPSHOT,$(CM_BUILDTYPE)),)
 PRODUCT_PROPERTY_OVERRIDES += \
